<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<!-- Global Site Tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-16521993-10"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-16521993-10');
</script>
	<title>The BGEN format</title>
	<link href="style.css" rel="stylesheet" type="text/css" />
	<script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
	<script type="text/javascript" src="js/local.js"></script>
</head>

<body>
  <div class="header">
    <div class="header_text">
      The BGEN format
    </div>

    <div class="header_subtext">
      A compressed binary format for typed and imputed genotype data
    </div>
	<!-- navigation -->
  </div>
	<nav>
		<ul>
			<li><a href="index.html">home</a></li>
			<li><a href="software.html">software</a></li>
			<li><a href="history.html">history</a></li>
			<li><a href="paper.html">paper</a></li>
			<li><a href="spec/latest.html">specification</a></li>
		</ul>
	</nav>
	<div id = "introduction" class="section">
		<h3>Introduction</h3>

		<p>
			Modern genetic association studies routinely employ data on tens to hundreds of thousands
			of individuals, genotyped or imputed at tens of millions of markers genome-wide.  Traditional
			data formats based on text representation of these data - such as
			the <a href="http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html">GEN</a>
			format output by <a href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html">IMPUTE</a>,
			or the <a href="http://ga4gh.org/#/fileformats-team">Variant Call Format</a>
			- are sometimes not well suited to these data quantities.  Indeed, for simple programs the time spent parsing
			these formats can dominate program execution time.
		</p>
		<p>
			This page describes a binary GEN file format (the "BGEN" format) which aims to address these problems.
			BGEN is a robust format that has been designed to have a specific blend of features that we believe make it
			useful for this type of study.  It is targetted for use with large, potentially imputed genetic datasets.
			Key features include:
			</ul>
				<li>The ability store both directly typed and imputed data.</li>
				<li>The ability to store both unphased genotypes and phased haplotype data.</li>
				<li>Small file sizes through the use of efficient, variable-precision packed bit representations and compression.</li>
				<li>The use of per-variant compression makes the format simple to index and easy to catalogue.</li>
			</ul>
		</p>
		<p>
			For example, the following plot shows the time taken to list variant identifying data - i.e. the genomic
			position, ID fields and alleles - for various common formats (Y-axis), against file size (X axis), for a dataset
			of 18,496 samples typed at 121,668 SNPs on chromosome 1.  Both variants of BGEN defined below are shown.
		</p>
		<div class="embedded_plot">
			<center>
				<img width = 400 height = 300 src="images/bgen_comparison.png"></img>
			</center>
		</div>
		<p>
			For <a href="https://www.cog-genomics.org/plink2/input#bed">PLINK binary (<code>.bed</code>) files</a>, identifying data is
			stored in a separate file (the <span class="monospace"><code>.bim</code></span> file) so the time is effectively zero.
			For text-based formats there is a significant trade-off between the use of file compression and
			read performance.  BGEN stores
			the entire dataset of 2,250 million genotypes in 334Mb, slightly over one bit per genotype, and in this test took 1.5s.
		</p>
		<p>
			(Performance optimisation of all formats may of course be possible, so the above plot
			will not represent the best possible timings, but should be regarded as illustrative.)
		</p>
		<p>
			The BGEN format has been used in several major projects, including the
			<a href="http://www.wtccc.org.uk/ccc2/">Wellcome Trust Case-Control Consortium 2</a>,
			the <a href="https://www.malariagen.net/projects/host">MalariaGEN</a> project, and the <a href="http://www.bristol.ac.uk/alspac/">ALSPAC study</a>.
			It has been adopted as the release format for genome-wide imputed genotypes
			for the <a href="http://www.ukbiobank.ac.uk">UK Biobank</a>.
		</p>
	</div>
	<div id = "contributors">
		<strong>Acknowledgements.</strong> The following people contributed to the design and implementation of the BGEN format:
        <ul>
                <li>
                    <a href="http://www.well.ox.ac.uk/~gav/">Gavin Band</a>
                </li>
                <li>
                    <a href="http://www.stats.ox.ac.uk/~marchini/">Jonathan Marchini</a>
                </li>
			</ul>
	</div>
</body>
</html>