<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<!-- Global Site Tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-16521993-10"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-16521993-10');
</script>
	<title>The BGEN format</title>
	<link href="style.css" rel="stylesheet" type="text/css" />
	<script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
	<script type="text/javascript" src="js/local.js"></script>
</head>


<body>
  <div class="header">
    <div class="header_text">
      The BGEN format
    </div>

    <div class="header_subtext">
      A compressed binary format for typed and imputed genotype data
    </div>
	<!-- navigation -->
  </div>
	<nav>
		<ul>
			<li><a href="index.html">home</a></li>
			<li><a href="software.html">software</a></li>
			<li><a href="history.html">history</a></li>
			<li><a href="paper.html">paper</a></li>
			<li><a href="spec/latest.html">specification</a></li>
		</ul>
	</nav>
	<div id = "tools">
		<h3>Software support</h3>
		<p>
			<b>For command-line users:</b> BGEN support has been implemented in several software packages, click on the links below for more information.
		</p>
		<div class="logo_container">
			<a href="http://www.well.ox.ac.uk/~gav/qctool"><span style="font-family: Ariel, Helvetica, Tahoma">qctool v2</span></a>
			<a href="http://mathgen.stats.ox.ac.uk/genetics_software/snptest/snptest.html"><code>SNPTEST</code></a>
			<a href="http://www.cog-genomics.org/plink2"><code>PLINK</code></a>
			<a href="http://watson.hgen.pitt.edu/mega2.html"><code>Mega2</code></a>
			<a href="http://www.christianbenner.com/"><img src="images/ldstore.png" alt="LDSTORE" height=20></a>
			<a href="http://www.geenivaramu.ee/en/tools/regscan"><img src="images/regscan.png" height=25 alt="REGSCAN"></a>
			<a href="http://jmarchini.org/bgenie/"><code>BGENIE</code></a>
			<a href="http://hail.is"><img src="images/hail-logo-cropped.png" alt = "Hail"  height=20></a>
			<a href="http://data.broadinstitute.org/alkesgroup/BOLT-LMM/"><code>BOLT-LMM</code></a>
			<a href="http://www.christianbenner.com/"><img src="images/finemap.png" height = 25 alt="FINEMAP"></a>
			<a href="https://github.com/rwdavies/STITCH"><code>STITCH</code></a>
			<a href="https://code.enkre.net/bgene.net/bgen/wiki/bgenix"><span style="font-family: Ariel, Helvetica, Tahoma">bgenix</span></a>
		</div>
		<p>
			<b>For R users:</b> The <a href="https://code.enkre.net/bgen/wiki/rbgen"><code>rbgen</code></a> package can be used to load data directly into R.
		</p>
		<p>

			<b>For python users:</b> A number of python solutions have been developed. These include the <a
			href="https://pypi.org/project/pysnptools/"><code>pysnptools</code></a> package, the <a
			href="https://pypi.org/project/bgen-reader">bgen-reader</a> library and the <a
			href="https://lemieuxl.github.io/pybgen/">pybgen</a> library.

		</p>
		<p>
			(Please contact me if your software supports BGEN and you'd like it added to this list.)
		</p>
		
		<p><b>For developers:</b> A freely available C++ implementation of the BGEN format is available at
			<a href="https://code.enkre.net/bgen">code.enkre.net</a>.  This repository also contains a number of useful tools:
		</p>
		<dl>
			<dt><a href="http://code.enkre.net/bgen/wiki/bgenix"><code>bgenix</code></a><dt>
				<dd>A tool which provides random access to data in a BGEN file by creating an index.</dd>
			<dt><a href="http://code.enkre.net/bgen/wiki/cat-bgen"><code>cat-bgen</code></a><dt>
				<dd>A program to efficiently concatenate BGEN files.  Together with <code>bgenix</code> this effectively
					enables map/reduce pipelines that operate in parallel on seperate genomic regions.
					For example, we have found this useful for joining data into chromosomes after imputing genome-wide.</dd>
				<dt><A href="https://code.enkre.net/bgen/wiki/edit-bgen"><code>edit-bgen</code></a></dt>
				<dd>A tool to (destructively) edit the BGEN header data.</dd>
		</dl>
		<p>
			In addition a
			set of <a href="https://code.enkre.net/bgen/src/release/example/">example files and programs</a>,
			intended to demonstrate the API, are provided.
		</p>

	</div>
</body>
</html>